internal/provider/resource_tfe_workspace_execution_mode.go

```diff
@@ -0,0 +1,186 @@
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		CustomizeDiff: validateExecutionMode,
+
+		Schema: map[string]*schema.Schema{
+			"agent_pool_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"workspace_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"execution_mode": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice(
+					[]string{
+						"agent",
+						"local",
+						"remote",
+					},
+					false,
+				),
+			},
+		},
+	}
+}
+
+func resourceTFEWorkspaceExecutionModeCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(ConfiguredClient)
+
+	executionMode := d.Get("execution_mode").(string)
+	agentPoolID := d.Get("agent_pool_id").(string)
+	workspaceID := d.Get("workspace_id").(string)
+
+	// Create a new options struct to attach the agent pool to workspace
+	options := tfe.WorkspaceUpdateOptions{
+		AgentPoolID:   tfe.String(agentPoolID),
+		ExecutionMode: tfe.String(executionMode),
+	}
+
+	log.Printf("[DEBUG] Create attachment on workspace with agent pool ID: %s", agentPoolID)
+	workspace, err := config.Client.Workspaces.UpdateByID(ctx, workspaceID, options)
+	if err != nil {
+		return fmt.Errorf("error attaching agent pool ID %s to workspace ID on CREATE %s: %w", agentPoolID, workspaceID, err)
+	}
+
+	d.SetId(workspace.ID)
+	// d.Set() will update the resource state file for execution_mode and agent pool ID
+	d.Set("execution_mode", workspace.ExecutionMode)
+	d.Set("agent_pool_id", workspace.AgentPool.ID)
+
+	return resourceTFEWorkspaceExecutionModeRead(d, meta)
+}
+
+func resourceTFEWorkspaceExecutionModeRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(ConfiguredClient)
+
+	log.Printf("[DEBUG] Read configuration: %s", d.Id())
+	workspace, err := config.Client.Workspaces.ReadByID(ctx, d.Id())
+	if err != nil {
+		if errors.Is(err, tfe.ErrResourceNotFound) {
+			log.Printf("[DEBUG] Workspace %s no longer exists", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("error reading configuration of workspace %s: %w", d.Id(), err)
+	}
+
+	d.Set("workspace_id", workspace.ID)
+
+	var poolID string
+	if workspace.AgentPool != nil {
+		poolID = workspace.AgentPool.ID
+	}
+	d.Set("agent_pool_id", poolID)
+
+	return nil
+}
+
+func resourceTFEWorkspaceExecutionModeUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(ConfiguredClient)
+
+	workspaceID := d.Get("workspace_id").(string)
+
+	options := tfe.WorkspaceUpdateOptions{}
+
+	if d.HasChange("execution_mode") {
+		options.ExecutionMode = tfe.String(d.Get("execution_mode").(string))
+	}
+
+	if d.HasChange("agent_pool_id") {
+		options.AgentPoolID = tfe.String(d.Get("agent_pool_id").(string))
+	}
+
+	log.Printf("[DEBUG] Update attachment on workspace with agent pool ID: %s", d.Get("agent_pool_id"))
```

Comment:
not all execution mode will be attaching agents, this should either be a generic message for all execution mode update or made conditional for agent_pool_id only